### ISSUE TITLE
Race condition in `syncx.RWMutex`.

### ISSUE THREAD

**Issue by jmalloc** [dogmatiq/infix#72]
```

••••••••••••••panic: close of nil channel

goroutine 16 [running]:
github.com/dogmatiq/infix/internal/x/syncx.(*RWMutex).RLock(0xc0000d9e80, 0x172ace0, 0xc000434380, 0x1, 0xc000051d98)
	/Users/james/grit/github.com/dogmatiq/infix/internal/x/syncx/mutex.go:120 +0x1c6
github.com/dogmatiq/infix/persistence/provider/memory.(*eventStoreResult).Next(0xc000434540, 0x172ace0, 0xc000434380, 0x0, 0xc00058a300, 0x0, 0x0)
	/Users/james/grit/github.com/dogmatiq/infix/persistence/provider/memory/eventstore.go:64 +0x8b
github.com/dogmatiq/infix/eventstream.(*eventStoreCursor).execQuery(0xc0000fe500, 0x172ace0, 0xc000434380, 0x0, 0x0)
	/Users/james/grit/github.com/dogmatiq/infix/eventstream/eventstore.go:183 +0x143
github.com/dogmatiq/infix/eventstream.(*eventStoreCursor).consume(0xc0000fe500, 0x172ace0, 0xc000434380)
	/Users/james/grit/github.com/dogmatiq/infix/eventstream/eventstore.go:164 +0x71
created by github.com/dogmatiq/infix/eventstream.(*EventStoreStream).Open
	/Users/james/grit/github.com/dogmatiq/infix/eventstream/eventstore.go:94 +0x2a6
FAIL	github.com/dogmatiq/infix/eventstream	1.113s
?   	github.com/dogmatiq/infix/eventstream/internal/streamtest	[no test files]
?   	github.com/dogmatiq/infix/fixtures	[no test files]
?   	github.com/dogmatiq/infix/internal/testing/boltdbtest	[no test files]
?   	github.com/dogmatiq/infix/internal/testing/sqltest	[no test files]
?   	github.com/dogmatiq/infix/internal/x/bboltx	[no test files]
?   	github.com/dogmatiq/infix/internal/x/grpcx	[no test files]
?   	github.com/dogmatiq/infix/internal/x/loggingx	[no test files]
?   	github.com/dogmatiq/infix/internal/x/sqlx	[no test files]
ok  	github.com/dogmatiq/infix/internal/x/syncx	(cached)
ok  	github.com/dogmatiq/infix/persistence	0.543s
?   	github.com/dogmatiq/infix/persistence/eventstore	[no test files]
ok  	github.com/dogmatiq/infix/persistence/provider/boltdb	1.238s
?   	github.com/dogmatiq/infix/persistence/provider/internal/providertest	[no test files]
ok  	github.com/dogmatiq/infix/persistence/provider/memory	0.202s
ok  	github.com/dogmatiq/infix/persistence/provider/sql	0.891s
?   	github.com/dogmatiq/infix/persistence/provider/sql/internal/query	[no test files]
ok  	github.com/dogmatiq/infix/persistence/provider/sql/mysql	2.236s
ok  	github.com/dogmatiq/infix/persistence/provider/sql/postgres	2.467s
ok  	github.com/dogmatiq/infix/persistence/provider/sql/sqlite	0.547s
ok  	github.com/dogmatiq/infix/projection	(cached)
ok  	github.com/dogmatiq/infix/projection/resource	(cached)

```

**Comment by jmalloc** [comment]
This occurred with the version of `RWMutex` in 1c0b646.

**Comment by jmalloc** [comment]
Ok, I *think* this occurs when there are two competing calls to `RLock()`, with an ordering something like:

I think this occurs when there are two competing calls to `RLock()`, where the winner completes its workload and unlocks the mutex *after* the loser unlocks the internal mutex but before it starts selecting.

The "loser" then enters the `<-m.unlocked` branch and attempts to close the channel that has already been closed, and set to `nil`, by the winner.

**Comment by jmalloc** [comment]
Reproduces fairly reliably with:

```go
			var g sync.WaitGroup

			for i := 0; i < 50; i++ {
				g.Add(1)
				go func() {
					defer g.Done()
					mutex.RLock(ctx)
					mutex.RUnlock()
				}()
			}

			g.Wait()
```

**Comment by jmalloc** [comment]
Fixed in #73.
